File: db4oj/core/src/com/db4o/reflect/generic/KnownClassesRepository.java
/* Copyright (C) 2006  db4objects Inc.  http://www.db4o.com */

package com.db4o.reflect.generic;

import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.marshall.*;
import com.db4o.reflect.*;

/**
 * @exclude
 */
public class KnownClassesRepository {
	
	// FIXME very java-centric, what about .NET?
	private final static Hashtable4 PRIMITIVES;
	
	static {
		PRIMITIVES=new Hashtable4();
		registerPrimitive(Boolean.class,boolean.class);
		registerPrimitive(Byte.class,byte.class);
		registerPrimitive(Short.class,short.class);
		registerPrimitive(Character.class,char.class);
		registerPrimitive(Integer.class,int.class);
		registerPrimitive(Long.class,long.class);
		registerPrimitive(Float.class,float.class);
		registerPrimitive(Double.class,double.class);
	}

	private static void registerPrimitive(Class wrapper,Class primitive) {
		PRIMITIVES.put(wrapper.getName(),primitive);
	}
	
	private ObjectContainerBase _stream;
	private Transaction _trans;

    private ReflectClassBuilder _builder;

    private final Hashtable4 _classByName = new Hashtable4();
    private final Hashtable4 _classByID = new Hashtable4();
	private Collection4 _pendingClasses = new Collection4();
    private final Collection4 _classes = new Collection4();

    public KnownClassesRepository(ReflectClassBuilder builder) {
    	_builder=builder;
    }
    
	public void setTransaction(Transaction trans){
		if(trans != null){
			_trans = trans;
			_stream = trans.container();
		}
	}
    
    public void register(ReflectClass clazz) {
    	register(clazz.getName(), clazz);
    }

    public ReflectClass forID(int id) {
    	synchronized(_stream.lock()) {
	    	if(_stream.handlers().isSystemHandler(id)) {
	    		return _stream.handlers().classForID(id);
	    	}
			ensureClassAvailability(id);
	        return lookupByID(id);        
    	}
    }
    
    public ReflectClass forName(String className) {
        final ReflectClass clazz = lookupByName(className);
        if(clazz != null){
            return clazz;
        }
        
        if(_stream == null) {
        	return null;
        }
        
		
    	synchronized(_stream.lock()) {
	        if(_stream.classCollection() == null){
	        	return null;
	        }
	        
	        int classID = _stream.classMetadataIdForName(className);
	        if(classID <= 0){
	        	return null;
	        }
	        
	        return initializeClass(classID, className);
    	}
    }

	private ReflectClass initializeClass(int classID, String className) {
		ReflectClass newClazz = ensureClassInitialised(classID);
        _classByName.put(className, newClazz);
        return newClazz;
	}

	private void readAll(){
		forEachClassId(new Procedure4<Integer>() { public void apply(Integer id) {
			ensureClassAvailability(id);
		}});
		forEachClassId(new Procedure4<Integer>() { public void apply(Integer id) {
			ensureClassRead(id);
		}});
	}
	
	private void forEachClassId(Procedure4<Integer> procedure) {
		for(Iterator4 ids=_stream.classCollection().ids();ids.moveNext();) {
			procedure.apply((Integer)ids.current());
		}
	}

	private ReflectClass ensureClassAvailability (int id) {

        if(id == 0){
            return null;
        }
		
        ReflectClass ret = (ReflectClass)_classByID.get(id);
		if(ret != null){
			return ret;
		}
        
		ByteArrayBuffer classreader=_stream.readWriterByID(_trans,id);

		ClassMarshaller marshaller = marshallerFamily()._class;
		RawClassSpec spec=marshaller.readSpec(_trans, classreader);

		String className = spec.name();
		ret = lookupByName(className);
		if(ret != null){
			_classByID.put(id, ret);
			_pendingClasses.add(new Integer(id));
			return ret;
		}
		
		ret = _builder.createClass(className, ensureClassAvailability(spec.superClassID()),spec.numFields());
		
		// step 1 only add to _classByID, keep the class out of _classByName and _classes
        _classByID.put(id, ret);
		_pendingClasses.add(new Integer(id));
		
		return ret;
	}

	private void ensureClassRead(int id) {

		ReflectClass clazz = lookupByID(id);
		
		ByteArrayBuffer classreader=_stream.readWriterByID(_trans,id);

		ClassMarshaller classMarshaller = marshallerFamily()._class;
		RawClassSpec classInfo=classMarshaller.readSpec(_trans, classreader);
		String className=classInfo.name();
		
		// Having the class in the _classByName Map for now indicates
		// that the class is fully read. This is breakable if we start
		// returning GenericClass'es in other methods like forName
		// even if a native class has not been found
		if(lookupByName(className) != null){
			return;
		}
		
        // step 2 add the class to _classByName and _classes to denote reading is completed
        register(className, clazz);
		
		int numFields=classInfo.numFields();
		ReflectField[] fields=_builder.fieldArray(numFields);
		FieldMarshaller fieldMarshaller=marshallerFamily()._field;
		
		for (int i = 0; i < numFields; i++) {
			
			RawFieldSpec fieldInfo=fieldMarshaller.readSpec(_stream, classreader);
			String fieldName=fieldInfo.name();
            ReflectClass fieldClass = reflectClassForFieldSpec(fieldInfo, _stream.reflector());
			fields[i]=_builder.createField(clazz, fieldName, fieldClass, fieldInfo.isVirtual(), fieldInfo.isPrimitive(), fieldInfo.isArray(), fieldInfo.isNArray());
		}
		_builder.initFields(clazz, fields);
	}

	private void register(String className, ReflectClass clazz) {
		if (lookupByName(className) != null)
			throw new IllegalArgumentException();
	    _classByName.put(className, clazz);
		_classes.add(clazz);
    }

	private ReflectClass reflectClassForFieldSpec(RawFieldSpec fieldInfo, Reflector reflector) {
		
		if (fieldInfo.isVirtual()) {
			 return virtualFieldByName(fieldInfo.name()).classReflector(reflector);
		}
		
		final int handlerID = fieldInfo.handlerID();
		
		// need to take care of special handlers here
		switch (handlerID){
		    case Handlers4.UNTYPED_ID:
		        return objectClass();
		    case Handlers4.ANY_ARRAY_ID:
		        return arrayClass(objectClass());
		    default:
		    	ReflectClass fieldClass=forID(handlerID);
		    	if (null != fieldClass) {
		    		return normalizeFieldClass(fieldInfo, fieldClass);
		    	}
		    	break;
		}
		return null;
	}

	private ReflectClass normalizeFieldClass(RawFieldSpec fieldInfo,
			ReflectClass fieldClass) {
		// TODO: why the following line is necessary?
		ReflectClass theClass=_stream.reflector().forName(fieldClass.getName());		    	
		if(fieldInfo.isPrimitive()) {
			theClass = primitiveClass(theClass);
		}
		if(fieldInfo.isArray()) {
			theClass = arrayClass(theClass);
		}
		return theClass;
	}

	private ReflectClass objectClass() {
		return _stream.reflector().forClass(Object.class);
	}

	private VirtualFieldMetadata virtualFieldByName(final String fieldName) {
		return _stream.handlers().virtualFieldByName(fieldName);
	}

	private MarshallerFamily marshallerFamily() {
		return MarshallerFamily.forConverterVersion(_stream.converterVersion());
	}

	private ReflectClass ensureClassInitialised (int id) {
		ReflectClass ret = ensureClassAvailability(id);
		while(_pendingClasses.size() > 0) {
			Collection4 pending = _pendingClasses;
			_pendingClasses = new Collection4();
			Iterator4 i = pending.iterator();
			while(i.moveNext()) {
				ensureClassRead(((Integer)i.current()).intValue());
			}
		}
		return ret;
	}
	
    public Iterator4 classes() {
    	readAll();
    	return _classes.iterator();
    }

    public void register(int id,ReflectClass clazz) {
    	_classByID.put(id, clazz);
    }
    
    public ReflectClass lookupByID(int id) {
    	return (ReflectClass)_classByID.get(id);
    }

    public ReflectClass lookupByName(String name) {
    	return (ReflectClass)_classByName.get(name);
    }
    
	private ReflectClass arrayClass(ReflectClass clazz) {
		Object proto=clazz.reflector().array().newInstance(clazz,0);
		return clazz.reflector().forObject(proto);
	}

	private ReflectClass primitiveClass(ReflectClass baseClass) {
		Class primitive=(Class) PRIMITIVES.get(baseClass.getName());
		if(primitive!=null) {
			return baseClass.reflector().forClass(primitive);
		}
		return baseClass;
	}
}

File: db4on/core/src/core/com/db4o/reflect/generic/KnownClassesRepository.cs
namespace com.db4o.reflect.generic
{
	public class KnownClassesRepository
	{
		private static readonly com.db4o.foundation.Hashtable4 PRIMITIVES;

		static KnownClassesRepository()
		{
			PRIMITIVES = new com.db4o.foundation.Hashtable4();
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(bool)), j4o.lang.JavaSystem.GetClassForType
				(typeof(bool)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(byte)), j4o.lang.JavaSystem.GetClassForType
				(typeof(byte)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(short)), j4o.lang.JavaSystem.GetClassForType
				(typeof(short)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(char)), j4o.lang.JavaSystem.GetClassForType
				(typeof(char)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(int)), j4o.lang.JavaSystem.GetClassForType
				(typeof(int)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(long)), j4o.lang.JavaSystem.GetClassForType
				(typeof(long)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(float)), j4o.lang.JavaSystem.GetClassForType
				(typeof(float)));
			RegisterPrimitive(j4o.lang.JavaSystem.GetClassForType(typeof(double)), j4o.lang.JavaSystem.GetClassForType
				(typeof(double)));
		}

		private static void RegisterPrimitive(j4o.lang.Class wrapper, j4o.lang.Class primitive
			)
		{
			PRIMITIVES.Put(wrapper.GetName(), primitive);
		}

		private com.db4o.@internal.ObjectContainerBase _stream;

		private com.db4o.@internal.Transaction _trans;

		private com.db4o.reflect.generic.ReflectClassBuilder _builder;

		private readonly com.db4o.foundation.Hashtable4 _classByName = new com.db4o.foundation.Hashtable4
			();

		private readonly com.db4o.foundation.Hashtable4 _classByID = new com.db4o.foundation.Hashtable4
			();

		private com.db4o.foundation.Collection4 _pendingClasses = new com.db4o.foundation.Collection4
			();

		private readonly com.db4o.foundation.Collection4 _classes = new com.db4o.foundation.Collection4
			();

		public KnownClassesRepository(com.db4o.reflect.generic.ReflectClassBuilder builder
			)
		{
			_builder = builder;
		}

		public virtual void SetTransaction(com.db4o.@internal.Transaction trans)
		{
			if (trans != null)
			{
				_trans = trans;
				_stream = trans.Stream();
			}
		}

		public virtual void Register(com.db4o.reflect.ReflectClass clazz)
		{
			_classByName.Put(clazz.GetName(), clazz);
			_classes.Add(clazz);
		}

		public virtual com.db4o.reflect.ReflectClass ForID(int id)
		{
			if (_stream.Handlers().IsSystemHandler(id))
			{
				return _stream.HandlerByID(id).ClassReflector();
			}
			EnsureClassAvailability(id);
			return LookupByID(id);
		}

		public virtual com.db4o.reflect.ReflectClass ForName(string className)
		{
			com.db4o.reflect.ReflectClass clazz = (com.db4o.reflect.ReflectClass)_classByName
				.Get(className);
			if (clazz != null)
			{
				return clazz;
			}
			if (_stream == null)
			{
				return null;
			}
			if (_stream.ClassCollection() != null)
			{
				int classID = _stream.ClassCollection().GetYapClassID(className);
				if (classID > 0)
				{
					clazz = EnsureClassInitialised(classID);
					_classByName.Put(className, clazz);
					return clazz;
				}
			}
			return null;
		}

		private void ReadAll()
		{
			for (System.Collections.IEnumerator idIter = _stream.ClassCollection().Ids(); idIter
				.MoveNext(); )
			{
				EnsureClassAvailability(((int)idIter.Current));
			}
			for (System.Collections.IEnumerator idIter = _stream.ClassCollection().Ids(); idIter
				.MoveNext(); )
			{
				EnsureClassRead(((int)idIter.Current));
			}
		}

		private com.db4o.reflect.ReflectClass EnsureClassAvailability(int id)
		{
			if (id == 0)
			{
				return null;
			}
			com.db4o.reflect.ReflectClass ret = (com.db4o.reflect.ReflectClass)_classByID.Get
				(id);
			if (ret != null)
			{
				return ret;
			}
			com.db4o.@internal.Buffer classreader = _stream.ReadWriterByID(_trans, id);
			com.db4o.@internal.marshall.ClassMarshaller marshaller = MarshallerFamily()._class;
			com.db4o.@internal.marshall.RawClassSpec spec = marshaller.ReadSpec(_trans, classreader
				);
			string className = spec.Name();
			ret = (com.db4o.reflect.ReflectClass)_classByName.Get(className);
			if (ret != null)
			{
				_classByID.Put(id, ret);
				_pendingClasses.Add(id);
				return ret;
			}
			ret = _builder.CreateClass(className, EnsureClassAvailability(spec.SuperClassID()
				), spec.NumFields());
			_classByID.Put(id, ret);
			_pendingClasses.Add(id);
			return ret;
		}

		private void EnsureClassRead(int id)
		{
			com.db4o.reflect.ReflectClass clazz = LookupByID(id);
			com.db4o.@internal.Buffer classreader = _stream.ReadWriterByID(_trans, id);
			com.db4o.@internal.marshall.ClassMarshaller classMarshaller = MarshallerFamily().
				_class;
			com.db4o.@internal.marshall.RawClassSpec classInfo = classMarshaller.ReadSpec(_trans
				, classreader);
			string className = classInfo.Name();
			if (_classByName.Get(className) != null)
			{
				return;
			}
			_classByName.Put(className, clazz);
			_classes.Add(clazz);
			int numFields = classInfo.NumFields();
			com.db4o.reflect.ReflectField[] fields = _builder.FieldArray(numFields);
			com.db4o.@internal.marshall.FieldMarshaller fieldMarshaller = MarshallerFamily().
				_field;
			for (int i = 0; i < numFields; i++)
			{
				com.db4o.@internal.marshall.RawFieldSpec fieldInfo = fieldMarshaller.ReadSpec(_stream
					, classreader);
				string fieldName = fieldInfo.Name();
				com.db4o.reflect.ReflectClass fieldClass = ReflectClassForFieldSpec(fieldInfo);
				fields[i] = _builder.CreateField(clazz, fieldName, fieldClass, fieldInfo.IsVirtual
					(), fieldInfo.IsPrimitive(), fieldInfo.IsArray(), fieldInfo.IsNArray());
			}
			_builder.InitFields(clazz, fields);
		}

		private com.db4o.reflect.ReflectClass ReflectClassForFieldSpec(com.db4o.@internal.marshall.RawFieldSpec
			 fieldInfo)
		{
			if (fieldInfo.IsVirtual())
			{
				com.db4o.@internal.VirtualFieldMetadata fieldMeta = _stream.Handlers().VirtualFieldByName
					(fieldInfo.Name());
				return fieldMeta.GetHandler().ClassReflector();
			}
			int handlerID = fieldInfo.HandlerID();
			com.db4o.reflect.ReflectClass fieldClass = null;
			switch (handlerID)
			{
				case com.db4o.@internal.HandlerRegistry.ANY_ID:
				{
					fieldClass = _stream.Reflector().ForClass(j4o.lang.JavaSystem.GetClassForType(typeof(object)
						));
					break;
				}

				case com.db4o.@internal.HandlerRegistry.ANY_ARRAY_ID:
				{
					fieldClass = ArrayClass(_stream.Reflector().ForClass(j4o.lang.JavaSystem.GetClassForType
						(typeof(object))));
					break;
				}

				default:
				{
					fieldClass = ForID(handlerID);
					fieldClass = _stream.Reflector().ForName(fieldClass.GetName());
					if (fieldInfo.IsPrimitive())
					{
						fieldClass = PrimitiveClass(fieldClass);
					}
					if (fieldInfo.IsArray())
					{
						fieldClass = ArrayClass(fieldClass);
					}
					break;
				}
			}
			return fieldClass;
		}

		private com.db4o.@internal.marshall.MarshallerFamily MarshallerFamily()
		{
			return com.db4o.@internal.marshall.MarshallerFamily.ForConverterVersion(_stream.ConverterVersion
				());
		}

		private com.db4o.reflect.ReflectClass EnsureClassInitialised(int id)
		{
			com.db4o.reflect.ReflectClass ret = EnsureClassAvailability(id);
			while (_pendingClasses.Size() > 0)
			{
				com.db4o.foundation.Collection4 pending = _pendingClasses;
				_pendingClasses = new com.db4o.foundation.Collection4();
				System.Collections.IEnumerator i = pending.GetEnumerator();
				while (i.MoveNext())
				{
					EnsureClassRead(((int)i.Current));
				}
			}
			return ret;
		}

		public virtual System.Collections.IEnumerator Classes()
		{
			ReadAll();
			return _classes.GetEnumerator();
		}

		public virtual void Register(int id, com.db4o.reflect.ReflectClass clazz)
		{
			_classByID.Put(id, clazz);
		}

		public virtual com.db4o.reflect.ReflectClass LookupByID(int id)
		{
			return (com.db4o.reflect.ReflectClass)_classByID.Get(id);
		}

		public virtual com.db4o.reflect.ReflectClass LookupByName(string name)
		{
			return (com.db4o.reflect.ReflectClass)_classByName.Get(name);
		}

		private com.db4o.reflect.ReflectClass ArrayClass(com.db4o.reflect.ReflectClass clazz
			)
		{
			object proto = clazz.Reflector().Array().NewInstance(clazz, 0);
			return clazz.Reflector().ForObject(proto);
		}

		private com.db4o.reflect.ReflectClass PrimitiveClass(com.db4o.reflect.ReflectClass
			 baseClass)
		{
			j4o.lang.Class primitive = (j4o.lang.Class)PRIMITIVES.Get(baseClass.GetName());
			if (primitive != null)
			{
				return baseClass.Reflector().ForClass(primitive);
			}
			return baseClass;
		}
	}
}
